cmd/juju/application/deployer/charm.go

```diff
@@ -458,53 +456,42 @@ func (c *repositoryCharm) compatibilityPrepareAndDeploy(ctx *cmd.Context, deploy
 		return errors.Trace(err)
 	}
 
-	var seriesFlag string
-	if !c.baseFlag.Empty() {
-		var err error
-		seriesFlag, err = corebase.GetSeriesFromBase(c.baseFlag)
-		if err != nil {
-			return errors.Trace(err)
-		}
-	}
-	supportedSeries, err := transform.SliceOrErr(supportedBases, corebase.GetSeriesFromBase)
+	imageStream := modelCfg.ImageStream()
+	workloadBases, err := SupportedJujuBases(c.clock.Now(), c.baseFlag, imageStream)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	selector := corecharm.SeriesSelector{
-		CharmURLSeries:      userRequestedURL.Series,
-		SeriesFlag:          seriesFlag,
-		SupportedSeries:     supportedSeries,
-		SupportedJujuSeries: workloadSeries,
+	baseSelector, err := corecharm.ConfigureBaseSelector(corecharm.SelectorConfig{
+		Config:              modelCfg,
 		Force:               c.force,
-		Conf:                modelCfg,
-		FromBundle:          false,
 		Logger:              logger,
+		RequestedBase:       c.baseFlag,
+		SupportedCharmBases: supportedBases,
+		WorkloadBases:       workloadBases,
 		UsingImageID:        (c.constraints.HasImageID() || c.modelConstraints.HasImageID()),
-	}
-	err = selector.Validate()
+	})
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	// Get the series to use.
-	series, err := selector.CharmSeries()
+	// Get the base to use.
+	base, err := baseSelector.CharmBase()
 
-	logger.Tracef("Using series %q from %v to deploy %v", series, supportedSeries, userRequestedURL)
+	logger.Tracef("Using base %q from %v to deploy %v", base, supportedBases, userRequestedURL)
 
-	imageStream := modelCfg.ImageStream()
 	// Avoid deploying charm if it's not valid for the model.
 	// We check this first before possibly suggesting --force.
 	if err == nil {
-		if err2 := c.validateCharmSeriesWithName(series, storeCharmOrBundleURL.Name, imageStream); err2 != nil {
+		if err2 := c.validateCharmBaseWithName(base, storeCharmOrBundleURL.Name, imageStream); err2 != nil {
 			return errors.Trace(err2)
 		}
 	}
 
 	if charm.IsUnsupportedSeriesError(err) {
```

Comment:
If SeriesSelector is removed, why do we still have a UnsupportedSeriesError type? Shouldn't we be checking for UnsupportedBaseError?

Reply:
Yes good spot. There's a place or two I noticed this, but missed this one clearly